Add Dockerfile instructions (#392)
AlexsLemonade/OpenScPCA-analysis: docs/software-platforms/docker/docker-images.md

# Creating Docker images

This page provides some guidance on creating Docker images for an OpenScPCA analysis module.

Note that creating docker images is not required for contributing to OpenScPCA, but it can be useful for testing and sharing your analysis module with others.
Each analysis module will ultimately be updated to include a Docker image to allow for running the analysis module as part of the `OpenScPCA-nf` workflow, but this does not need to be done by the module author.
Data Lab staff will create a Docker image as part of the process of adding the module to the `OpenScPCA-nf` workflow if it has not already been done.

## Prerequisites

Before making a Docker image, you will likely want to have fully [defined the software environment](../../contributing-to-analyses/determining-requirements/determining-software-requirements.md) that you will be using for your analysis module.
Most likely, this will take the form of either:

> - an [`renv.lock` file](../../contributing-to-analyses/determining-requirements/determining-software-requirements.md#determining-and-managing-software-dependencies-in-r) if you are working with R, or

Reviewer: "an" 🌶️

- a [conda `environment.yml`](../../contributing-to-analyses/determining-requirements/determining-software-requirements.md#module-specific-conda-environments) and/or [`conda-lock.yml` file](../../contributing-to-analyses/determining-requirements/determining-software-requirements.md#freezing-dependencies-with-conda-lock) if you are working with Python or other software available through [Bioconda](https://bioconda.github.io).

> The goal of creating a Docker image is to incorporate these software requirements into a container that can be used to run all parts of the analysis module without any installation steps other than installing Docker and pulling the image.

Reviewer: I'd move this up to the top of the document, seems like important context for the page overall.

Reviewer: it's kind of long too, so maybe 2 sentence-ify it

Author: I kind of reworked this to have a different emphasis, and moved it.

> You will also need to have [Docker installed on your computer](https://docs.docker.com/engine/install/) to build and test the images.

Reviewer: fyi we have this! https://openscpca.readthedocs.io/en/latest/technical-setup/environment-setup/install-docker/

Author: Changing the link!

## Docker template files

> When you create an analysis module, a template `Dockerfile` and a `.dockerignore` file will be created in the root of the analysis module.

Reviewer: I'd link to the analysis modules index page here

> The  `Dockerfile` includes a `FROM` statement that specifies the base image to use and a couple of `LABEL` statements to populate metadata about the image.

Reviewer: ```suggestion
The `Dockerfile` includes a `FROM` statement that specifies the base image to use and a couple of `LABEL` statements to populate metadata about the image.
```

The `.dockerignore` file is used to specify files and directories that should *not* be included in the Docker build environment.

> By default, this will be set to ignore all files with the exception of environment files (e.g., `renv.lock`, `environment.yml`, `conda-lock.yml`).

Reviewer: ```suggestion
By default, this will be set to ignore all files except environment files (e.g., `renv.lock`, `environment.yml`, `conda-lock.yml`).
```

This will keep the build environment small and prevent unnecessary files from being included in the image.

## Base images

The base image you choose will depend on the software requirements of your analysis module.
You should aim to use a publicly available base image that is in active development and well-maintained in order to ensure that your image is secure and up-to-date.
You should always use a *versioned* base image to ensure that your analysis module is reproducible, i.e., not the one tagged as `latest`.

## Additional software

If your analysis module requires additional software that is not available in the base image, you will need to install it in the Dockerfile.
The easiest way to do this is likely to be to use the same tool (`renv`, `conda`, etc.) that you used to define the software environment for your analysis module.
If necessary, you can also include statements to install other software using the system package manager (e.g., `apt-get` for Debian-based images).

## R-based images

> If you are working with R, you will likely want to use one of the [`rocker/r-ver` images](https://hub.docker.com/r/rocker/r-ver/tags) or a[`bioconductor/r-ver` image](https://hub.docker.com/r/bioconductor/r-ver/tags) images, which are a set of Docker images that provide complete R environments, tied to a specific version of R and/or Bioconductor.

Reviewer: ```suggestion
If you are working with R, you will likely want to use one of the [`rocker/r-ver` images](https://hub.docker.com/r/rocker/r-ver/tags) or a [`bioconductor/r-ver` image](https://hub.docker.com/r/bioconductor/r-ver/tags) images, which are a set of Docker images that provide complete R environments, tied to a specific version of R and/or Bioconductor.
```

> The Bioconductor images are especially handy as they include system software dependencies for Bioconductor packages that may not be installed in the `rocker` images, and support binary installation of Bioconductor packages, which can dramatically speed up build times.

Reviewer: ```suggestion
The Bioconductor images are especially handy as they include system software dependencies for Bioconductor packages that may not be installed in the `rocker` images.
They also support binary installation of Bioconductor packages, which can dramatically speed up build times.
```

You will want to include the following steps in your Dockerfile:

- Install `renv` (and `remotes` if you have any GitHub-based packages in your environment).
- Copy the `renv.lock` file from the host environment to the image.
- Use `renv::restore()` to reproduce the R environment from the `renv.lock` file.

> A simple `Dockerfile` for an R-based analysis module that uses `renv` for its environment might look like this:

Reviewer: what about `LABEL`? if it's in the template, it should be here too

Author: I wanted to focus on the mechanics with these examples. The labels are long and distracting.

> A simple `Dockerfile` for an R-based analysis module that uses `renv` for its environment might look like this:

Reviewer: ```suggestion
A simple `Dockerfile` for an R-based analysis module that uses `renv` for its environment might begin like this:
```

ENV RENV_CONFIG_CACHE_ENABLED FALSE

# Copy the renv.lock file from the host environment to the image
COPY renv.lock renv.lock

# restore from renv.lock file and clean up to reduce image size
RUN Rscript -e 'renv::restore()' && \
  rm -rf ~/.cache/R/renv && \
  rm -rf /tmp/downloaded_packages && \
  rm -rf /tmp/Rtmp*
```

## Conda-based images

If you are working with conda, you will likely want to use one of the [`continuumio/miniconda3` images](https://hub.docker.com/r/continuumio/miniconda3/tags).
Since the conda environment files should define the version of Python itself, you should not need to worry too much about the specific version of the base image, but you should still use a versioned image for reproducibility.

In the Dockerfile, you will want include the following steps:

- Install `conda-lock`.
- Copy the `conda-lock.yml` file from the host environment to the image.
- Use `conda-lock install` to create an environment from the `conda-lock.yml` file.
- Make sure the environment is activated when the container is launched by modifying the `.bashrc` file and setting the `ENTRYPOINT` to `bash -l -c`.

> A simple `Dockerfile` for a conda-based analysis module that uses `conda-lock` for its environment might look like this:

Reviewer: ```suggestion
A simple `Dockerfile` for a conda-based analysis module that uses `conda-lock` for its environment might begin like this:
```

Author: This is the whole thing though? (Aside from labels, as noted)

Reviewer: Ah yeah, I suppose it is. I was thinking about what if more dependencies are needed, but they'd be handled with the conda lines!

> ```Dockerfile

Reviewer: same `LABEL`  comment

# Creating Docker images

This page provides some guidance on creating Docker images for an OpenScPCA analysis module.

> Having a Docker image for your analysis module is a key part of ensuring that your analysis is as easily reproducible as possible.

Reviewer 2: ```suggestion
Having a Docker image for your analysis module is a key part of ensuring that your analysis is as reproducible as possible.
```

You will also need to have [Docker installed on your computer](../../technical-setup/environment-setup/install-docker.md) to build and test the images.

## Docker template files

When you [create an analysis module](../../contributing-to-analyses/analysis-modules/creating-a-module.md), a template `Dockerfile` and a `.dockerignore` file will be created in the root of the analysis module.

The `Dockerfile` includes a `FROM` statement that specifies the base image to use and a couple of `LABEL` statements to populate metadata about the image.

The `.dockerignore` file is used to specify files and directories that should *not* be included in the Docker build environment.
By default, this will be set to ignore all files except environment files (e.g., `renv.lock`, `environment.yml`, `conda-lock.yml`).
This will keep the build environment small and prevent unnecessary files from being included in the image.

## Base images

The base image you choose will depend on the software requirements of your analysis module.
You should aim to use a publicly available base image that is in active development and well-maintained in order to ensure that your image is secure and up-to-date.

> You should always use a *versioned* base image to ensure that your analysis module is reproducible, i.e., not the one tagged as `latest`.

Reviewer 2: Maybe just add a note here that we provide recommendations on base images to use for R and python below?

# Creating Docker images

This page provides some guidance on creating Docker images for an OpenScPCA analysis module.

Having a Docker image for your analysis module is a key part of ensuring that your analysis is as reproducible as possible.

Note that creating docker images is not required for contributing to OpenScPCA, but it can be useful for testing and sharing your analysis module with others.
Each analysis module will ultimately be updated to include a Docker image before producing final results, but this does not need to be done by the module author.

> Data Lab staff will create a Docker image as part of the process of adding the module to the `OpenScPCA-nf` workflow if it has not already been done.

Author: We are going to need to add more references to it at some point in the not too distant future. Once we have some downstream analysis that we want available to users, we will want to show people where that comes from.

> Data Lab staff will create a Docker image as part of the process of adding the module to the `OpenScPCA-nf` workflow if it has not already been done.

Reviewer: Ah I see it also in `docs/software-platforms/docker/index.md`, but without much explanation there either.

## Docker template files

When you [create an analysis module](../../contributing-to-analyses/analysis-modules/creating-a-module.md), a template `Dockerfile` and a `.dockerignore` file will be created in the root of the analysis module.

The `Dockerfile` includes a `FROM` statement that specifies the base image to use and a couple of `LABEL` statements to populate metadata about the image.

The `.dockerignore` file is used to specify files and directories that should *not* be included in the Docker build environment.
By default, this will be set to ignore all files except environment files (e.g., `renv.lock`, `environment.yml`, `conda-lock.yml`).
This will keep the build environment small and prevent unnecessary files from being included in the image.

## Base images

The base image you choose will depend on the software requirements of your analysis module.
You should aim to use a publicly available base image that is in active development and well-maintained in order to ensure that your image is secure and up-to-date.
You should always use a *versioned* base image to ensure that your analysis module is reproducible, i.e., not the one tagged as `latest`.

> We present some recommended base images for R and conda-based environments below, but you may need to choose a different base image depending on your specific requirements.

Reviewer: Something about this organization seemed a smidge off to me, so I suggested some change in headers below, but I'm not entirely sure how I feel about the suggestions. I feel like R/conda-based images should be nested within a base images header, so you could also move those two as `###` sections up here, and then keep `## additional software` as the last header

Author: > Something about this organization seemed a smidge off to me, so I suggested some change in headers below, but I'm not entirely sure how I feel about the suggestions. I feel like R/conda-based images should be nested within a base images header, so you could also move those two as `###` sections up here, and then keep `## additional software` as the last header

I updated the organization a little bit differently, making two main headers and putting these under an "Examples" header. Let me know what you think

Reviewer: I like this a lot!

## Additional software

If your analysis module requires additional software that is not available in the base image, you will need to install it in the Dockerfile.
The easiest way to do this is likely to be to use the same tool (`renv`, `conda`, etc.) that you used to define the software environment for your analysis module.
If necessary, you can also include statements to install other software using the system package manager (e.g., `apt-get` for Debian-based images).

> 

Reviewer: ```suggestion

## Base image recommendations

```

They also support binary installation of Bioconductor packages, which can dramatically speed up build times.

You will want to include the following steps in your Dockerfile:

- Install `renv` (and `remotes` if you have any GitHub-based packages in your environment).
- Copy the `renv.lock` file from the host environment to the image.
- Use `renv::restore()` to reproduce the R environment from the `renv.lock` file.

A simple `Dockerfile` for an R-based analysis module that uses `renv` for its environment might look like this:

```Dockerfile
# Base image on the Bioconductor 3.18 image
FROM bioconductor/r-ver:3.18

# Install renv to enable later package installation
RUN Rscript -e "install.packages('renv')"

# Disable the renv cache to install packages directly into the R library
ENV RENV_CONFIG_CACHE_ENABLED FALSE

# Copy the renv.lock file from the host environment to the image
COPY renv.lock renv.lock

# restore from renv.lock file and clean up to reduce image size
RUN Rscript -e 'renv::restore()' && \
  rm -rf ~/.cache/R/renv && \
  rm -rf /tmp/downloaded_packages && \
  rm -rf /tmp/Rtmp*
```

> ## Conda-based images

Reviewer: ```suggestion
### Conda-based images
```
